src/installer/installer.go

```diff
@@ -167,10 +168,25 @@ func (i *installer) InstallNode() error {
 	}
 	//update installation progress
 	i.UpdateHostInstallProgress(models.HostStageRebooting, "")
-	//reboot
-	if err = i.ops.Reboot(); err != nil {
+	// if ironic-agent exists on the host stop the assisted-agent service and exit, else reboot
+	ironicAgentServiceName := "ironic-agent.service"
+
+	out, err := i.ops.ExecPrivilegeCommand(nil, "systemctl", "list-units", "--no-legend", ironicAgentServiceName)
+	if err != nil {
+		i.log.Errorf("Failed to check if ironic agent service exists on the node %s", err)
 		return err
 	}
+	if strings.Contains(out, ironicAgentServiceName) {
```

Comment:
Can we move this logic to the systemd unit file level, and have this logic be a simple configuration at the agent level? Something like "AVOID_REBOOT: true, SELF_SERVICE_NAME: agent.service"

Reply:
Having very specific logic like this makes it harder to run the agent in "unnatural" environments, like the swarm.

Reply:
This means someone else will need to set this configuration.
With the current implementation, the only place that cares about the ironic service existence and how to handle it is here. 
Side note, the assisted-installer isn't a service, in case ironic agent also exists on the host the installer will stop the assisted service agent service: `agent.service`

Reply:
>This means someone else will need to set this configuration.

Exactly, I want the unit file that runs the agent to:

1. Check if there's an ironic service defined in the system
2. If yes, configure the agent to AVOID_REBOOT
3. Tell the agent its own service name through SELF_SERVICE_NAME, set it to `agent.service`

But when you're putting all this logic inside the agent itself, you're assuming the agent:
1. Runs in a systemd environment
2. Assuming the agent itself runs as a systemd service
3. Assuming the agent runs in a systemd service called "agent.service"

All these assumptions make it harder to run the agent in different environments - for example, as a library, like @ori-amizur  is working on

Reply:
>Side note, the assisted-installer isn't a service, in case ironic agent also exists on the host the installer will stop the assisted service agent service: agent.service

Not sure I understand this part, can you elaborate?

Reply:
> Not sure I understand this part, can you elaborate?
Yes, and I think it will explain why I prefer putting it all here.
This code isn't the agent.service, this code get compiled and shipped as the assisted-installer image.
The assisted-installer image get executed (podman run... ) when the assisted-agent get the install command.

Reply:
:man_facepalming: ugh I got confused and then thought we're in the agent repo lol.

Reply:
> logic like this makes it harder to run the agent in "unnatural" environments, like the swarm.
So I can add:
```
if i.DryRunEnabled {
		return nil
	}
```
As we do for all other systemctl operations

Reply:
My point still stands though - the agent should be configured through the systemd unit to know its own systemd service name, and it should be configured to *tell the installer* (through the installer CLI args) to avoid a reboot and close the agent instead

Reply:
What will we gain from it?
Except from splitting the logic to 2 different code bases?
Why does the agent service needs to care about the ironic agent?
It's not different than saying the assisted-agent should check the hostname and pass it in a cli option instead of checking for it in the installer code

Reply:
hmm, can't it be done from service side? i mean don't we have some configuration saying we are running with this specific mode? 
I am not sure regarding stopping agent in installer as installer is ran from agent , sort of killing your own father?
As installer command is known to agent we can add params to it easy from agent itself and then all logic will be in agent and i think this logic should be part of agent and not installer.

Reply:
We pass a flag from from the service side, we can also add API for the user to control it...
But why would we want to?
The only place that cares about and knows how to perform the installation is here.